Update Course: Form refactoring, including datetime-picker

thilo-behnke opened the pull request:
<!-- Thanks for contributing to ArTEMiS! Before you submit your pull request, please make sure to check the following boxes by putting an x in the [ ] (don't: [x ], [ x], do: [x]) -->

### Checklist
- [x] I've run `yarn run webpack:build:main` from the root directory to see that the project builds without errors.
- [x] I've removed unnecessary whitespace changes.
- [ ] I've updated the documentation and models if necessary.
- [ ] I've tested the changes and all related features on the Artemis test server.
- [ ] I've added (end-to-end) test cases for all new functionality.

### Open Issues
~~I had issues running the tests on my computer, I will look into this tomorrow.~~
I would also like to add some tests for date-time-picker.

**1) All e2e tests are failing for me**
I run _yarn test:e2e_, but all tests fail.
I see that the browser opens a new session on localhost:8080 and accesses artemis, but it seems to me that the tests can't login. Is this a known issue? Or am I missing something?

Here an example error:
  1) account
       should fail to login with bad password:
     TypeError: each key must be a number of string; got undefined
     [...]
      at SignInPage.setUserName (src/test/javascript/e2e/page-objects/jhi-page-objects.ts:280:24)

**2) Unit/Integration Tests not running**
I run _yarn test_ which triggers jest.
However jest is not a dependency of this project.
When I add jest myself and run yarn test again, I get this error:

Error: Could not resolve a module for a custom reporter.
  Module name: jest-junit

**3) Travis build error**
Link: https://api.travis-ci.com/v3/job/178294900/log.txt
Travis seems to have an issue when installing a new dependency I added:

Unexpected value 'OwlMomentDateTimeModule in /home/travis/build/ls1intum/ArTEMiS/node_modules/**ng-pick-datetime-moment**/moment-adapter/moment-date-time.module.d.ts' imported by the module 'FormDateTimePickerModule in /home/travis/build/ls1intum/ArTEMiS/src/main/webapp/app/shared/dateTimePicker/date-time-picker.module.ts'. Please add a @NgModule annotation.

For me locally I don't have this issue, I made sure that I have no uncommited changes.

### Motivation and Context
I noticed potential for enhancement in the update course form concerning usability and visualization.

### Description
**1) I refactored the form from a template into a reactive form.**
The reason for this was that form validation is more straight forward.
Form validation is currently only showing messages for the short name field, as its requirements are most complicated (I'm not showing messages for required fields, that should be clear from the bootstrap styling)
**2) I refactored the date-picker component into an inline styled one (i saw the todo in the code).**
I had to implement the ControlValueAccess interface so that forms can interact with the component without any tricks (e.g. supplying object and key for value binding).
I went through the code and changed the places where the component was used to make sure everything still works.
**3) I aligned date fields next to each other in a flex container**, because they shouldn't need too much space. I also added a calendar icon for toggling.

### Some further improvements that could be done
- Add cross validation on course start and course end fields in course-update (start needs to be before end, etc)
- Add validation for date in start/end (can not be in past, etc.)

#### Screenshots

![image](https://user-images.githubusercontent.com/28230611/52892765-fb72be80-3195-11e9-9ad4-f3b4508a78a3.png)

![image](https://user-images.githubusercontent.com/28230611/52892773-09c0da80-3196-11e9-9038-ef72883bc94d.png)

krusche commented:
Thanks for this contribution. It looks good so far. First we need to fix the build issue: Did you invoke `yarn webpack:build` and `yarn webpack:prod` locally? It is important that the build works, otherwise we cannot merge it

krusche commented on package.json:
```diff
@@ -39,6 +39,7 @@
         "moment": "2.24.0",
         "ng-jhipster": "0.9.3",
         "ng-pick-datetime": "7.0.0",
+        "ng-pick-datetime-moment": "^1.0.8",
```
Please do not use `^` as it can lead to build issues when the dependency is updated

thilo-behnke replied:
Thanks, you are right!

thilo-behnke commented:
Thanks for the review!

What I did in my last commits:
- Fixed issue with build (I had to remove the datetime moment format dependency, as it is not working well with angular 7 atm. I added a comment in the code for future reference)
- Fixed the unit tests for course update
- Moved the calendar icon with the input field, so the user doesn't have to move the mouse as far:

![image](https://user-images.githubusercontent.com/28230611/52946303-06834580-3374-11e9-95d9-3777acf8ec45.png)

thilo-behnke commented:
This PR was migrated to: https://github.com/ls1intum/ArTEMiS/pull/168.
